graph.go

```diff
@@ -0,0 +1,288 @@
+	v1 *vertex
+	v2 *vertex
+}
+
+type cycle []*vertex
+type cycleList []cycle
+
+func (g *Graph) Target() *Spec {
+	return g.specs[g.target]
+}
+
+func (g *Graph) Get(name string) (*Spec, bool) {
+	s, ok := g.specs[name]
+	return s, ok
+}
+
+// OrderedSlice returns an array of Specs in dependency order, up to and
+// including `target`. If `target` is the empty string, return the entire list.
+// If the target is not found, return an empty array.
+func (g *Graph) OrderedSlice(target string) []*Spec {
+	if target == "" {
+		return []*Spec(g.ordered)
+	}
+
+	for i, dep := range g.ordered {
+		if dep.Name == target {
+			return g.ordered[:i+1]
+		}
+	}
+
+	return []*Spec{}
+}
+
+// Returns the length of the ordred list of dependencies, up to and including
+// `target`. If `target` is the empty string, return the entire length.
+func (g *Graph) OrderedLen(target string) int {
+	if target == "" {
+		return len(g.ordered)
+	}
+
+	return len(g.OrderedSlice(target))
+}
+
+type vertex struct {
+	name    string
+	index   *int
+	lowlink int
+	onStack bool
+}
+
+var (
+	graphLock  sync.Mutex
+	BuildGraph *Graph
+	NotFound   = errors.New("dependency not found")
+)
+
+func (g *Graph) Last() *Spec {
+	return g.ordered[len(g.ordered)-1]
+}
+
+func (g *Graph) Lock() {
+	graphLock.Lock()
+}
+func (g *Graph) Unlock() {
+	graphLock.Unlock()
+}
+
+func InitGraph(specs []*Spec, subTarget string) error {
+	if BuildGraph != nil {
+		return nil
+	}
+
+	if BuildGraph == nil {
+		BuildGraph = new(Graph)
+		BuildGraph.Lock()
+		defer BuildGraph.Unlock()
+		*BuildGraph = Graph{
+			target:   subTarget,
+			edges:    sets.New[dependency](),
+			vertices: make([]*vertex, len(specs)),
+			specs:    make(map[string]*Spec),
+			indices:  make(map[string]int),
+			ordered:  nil,
+		}
+	}
+
+	if len(specs) == 1 {
+		BuildGraph.target = specs[0].Name
+	}
+
+	for i, spec := range specs {
+		name := spec.Name
+		BuildGraph.specs[name] = spec
+		v := &vertex{name: name}
+		BuildGraph.indices[name] = i
+		BuildGraph.vertices[i] = v
+	}
+
+	for name, spec := range BuildGraph.specs {
+		if spec.Dependencies == nil {
+			continue
+		}
+		vi := BuildGraph.indices[name]
+		v := BuildGraph.vertices[vi]
+		type depMap map[string][]string
+		runtimeAndBuildDeps := []depMap{
+			spec.Dependencies.Build,
+			spec.Dependencies.Runtime,
+		}
+
+		for _, deps := range runtimeAndBuildDeps {
+			if deps == nil {
+				continue
+			}
+
+			for dep, constraints := range deps {
+				_ = constraints // TODO(pmengelbert)
+				if name == dep {
+					continue // ignore if cycle is length 1
+				}
+				wi, ok := BuildGraph.indices[dep]
+				if !ok {
+					// this is dependency in the package repo
+					continue
+				}
+				w := BuildGraph.vertices[wi]
+				BuildGraph.edges.Insert(dependency{
+					v1: v,
+					v2: w,
+				})
+			}
+		}
+	}
+
+	return BuildGraph.topSort()
+}
+
+// https://en.wikipedia.org/wiki/Tarjan%27s_strongly_connected_components_algorithm
+func (g *Graph) topSort() error {
```

Comment:
can we split this function up? have one function that finds strongly connected components, and another that checks to make sure that components have only one vertex each? (i.e., L232 on in a separate function)

Reply:
Sure

Reply:
done